# 📌 Extraction


In [22]:
import pandas as pd
import json

url = 'https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json'

In [29]:
df = pd.read_json(url)
df.sample(10)

customerID Churn                                           customer  \
709   1017-FBQMM    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
234   0354-WYROK   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
4952  6765-MBQNU    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
211   0322-CHQRU   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
5052  6904-JLBGY    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
5961  8148-WOCMK    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3157  4398-HSCJH   Yes  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
63    0106-UGRDO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
3241  4525-VZCZG    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
6516  8957-THMOA    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   

                                                phone  \
709   {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
234    {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4952  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
211    {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
5052  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
5961   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3157  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
63    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
3241   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
6516   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                               internet  \
709   {'InternetService': 'Fiber optic', 'OnlineSecu...   
234   {'InternetService': 'Fiber optic', 'OnlineSecu...   
4952  {'InternetService': 'No', 'OnlineSecurity': 'N...   
211   {'InternetService': 'DSL', 'OnlineSecurity': '...   
5052  {'InternetService': 'Fiber optic', 'OnlineSecu...   
5961  {'InternetService': 'No', 'OnlineSecurity': 'N...   
3157  {'InternetService': 'DSL', 'OnlineSecurity': '...   
63    {'InternetService': 'Fiber optic', 'OnlineSecu...   
3241  {'InternetService': 'Fiber optic', 'OnlineSecu...   
6516  {'InternetService': 'No', 'OnlineSecurity': 'N...   

                                                account  
709   {'Contract': 'One year', 'PaperlessBilling': '...  
234   {'Contract': 'Month-to-month', 'PaperlessBilli...  
4952  {'Contract': 'One year', 'PaperlessBilling': '...  
211   {'Contract': 'Month-to-month', 'PaperlessBilli...  
5052  {'Contract': 'Two year', 'PaperlessBilling': '...  
5961  {'Contract': 'Two year', 'PaperlessBilling': '...  
3157  {'Contract': 'Month-to-month', 'PaperlessBilli...  
63    {'Contract': 'Two year', 'PaperlessBilling': '...  
3241  {'Contract': 'One year', 'PaperlessBilling': '...  
6516  {'Contract': 'Two year', 'PaperlessBilling': '...

# 🔧 Transform

In [30]:
df_dict = df.to_dict(orient="records")
df = pd.json_normalize(df_dict, sep='_')
df.sample(10)

customerID Churn customer_gender  customer_SeniorCitizen  \
2720  3766-EJLFL    No          Female                       0   
2954  4102-HLENU    No          Female                       0   
593   0843-WTBXE   Yes            Male                       0   
1220  1741-WTPON    No            Male                       0   
1780  2525-GVKQU    No          Female                       0   
1431  2057-BOYKM    No          Female                       1   
1836  2612-RANWT   Yes          Female                       0   
3466  4803-LBYPN    No            Male                       0   
2919  4060-BTPXO                Female                       0   
3374  4686-GEFRM    No            Male                       0   

     customer_Partner customer_Dependents  customer_tenure phone_PhoneService  \
2720              Yes                 Yes               68                Yes   
2954              Yes                  No               67                Yes   
593                No                  No               53                Yes   
1220              Yes                 Yes               56                Yes   
1780               No                  No               60                Yes   
1431              Yes                  No               72                Yes   
1836               No                  No               12                Yes   
3466              Yes                 Yes               34                Yes   
2919              Yes                  No               67                Yes   
3374              Yes                  No               70                Yes   

     phone_MultipleLines internet_InternetService  ... internet_OnlineBackup  \
2720                 Yes              Fiber optic  ...                    No   
2954                 Yes                      DSL  ...                   Yes   
593                  Yes              Fiber optic  ...                    No   
1220                 Yes              Fiber optic  ...                   Yes   
1780                  No                       No  ...   No internet service   
1431                 Yes              Fiber optic  ...                   Yes   
1836                 Yes              Fiber optic  ...                    No   
3466                  No                       No  ...   No internet service   
2919                 Yes              Fiber optic  ...                   Yes   
3374                 Yes              Fiber optic  ...                   Yes   

     internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
2720                       Yes                  Yes                  Yes   
2954                       Yes                   No                   No   
593                         No                   No                  Yes   
1220                       Yes                  Yes                  Yes   
1780       No internet service  No internet service  No internet service   
1431                       Yes                  Yes                   No   
1836                       Yes                   No                  Yes   
3466       No internet service  No internet service  No internet service   
2919                       Yes                   No                  Yes   
3374                       Yes                  Yes                   No   

     internet_StreamingMovies account_Contract account_PaperlessBilling  \
2720                      Yes         Two year                      Yes   
2954                       No         Two year                       No   
593                        No   Month-to-month                       No   
1220                      Yes         Two year                       No   
1780      No internet service         Two year                       No   
1431                       No         Two year                      Yes   
1836                      Yes   Month-to-month                      Yes   
3466      No internet service   Month-to-month                       

#### Data dictionary

* `customerID`: unique identification number for each customer
* `Churn`: whether or not the customer left the company
* `gender`: gender (male and female)
* `SeniorCitizen`: whether the customer is 65 years old or older
* `Partner`: whether or not the customer has a partner
* `Dependents`: whether or not the customer has dependents
* `tenure`: number of months the customer has been under contract
* `PhoneService`: subscription to phone service
* `MultipleLines`: subscription to more than one phone line
* `InternetService`: subscription to an internet service provider
* `OnlineSecurity`: additional online security subscription
* `OnlineBackup`: additional online backup subscription
* `DeviceProtection`: additional device protection subscription
* `TechSupport`: additional technical support subscription with shorter waiting times
* `StreamingTV`: cable TV subscription
* `StreamingMovies`: movie streaming subscription
* `Contract`: type of contract
* `PaperlessBilling`: whether the customer prefers to receive bills online
* `PaymentMethod`: payment method
* `Charges.Monthly`: total cost of all customer services per month
* `Charges.Total`: total amount spent by the customer


In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer_gender            7267 non-null   object 
 3   customer_SeniorCitizen     7267 non-null   int64  
 4   customer_Partner           7267 non-null   object 
 5   customer_Dependents        7267 non-null   object 
 6   customer_tenure            7267 non-null   int64  
 7   phone_PhoneService         7267 non-null   object 
 8   phone_MultipleLines        7267 non-null   object 
 9   internet_InternetService   7267 non-null   object 
 10  internet_OnlineSecurity    7267 non-null   object 
 11  internet_OnlineBackup      7267 non-null   object 
 12  internet_DeviceProtection  7267 non-null   object 
 13  internet_TechSupport       7267 non-null   objec

In [84]:
df = df.apply(lambda x: x.str.strip() if x.dtype == "object" else x)
[(df == '').sum(), df.isna().sum()]

[customerID                     0
 Churn                        224
 customer_gender                0
 customer_SeniorCitizen         0
 customer_Partner               0
 customer_Dependents            0
 customer_tenure                0
 phone_PhoneService             0
 phone_MultipleLines            0
 internet_InternetService       0
 internet_OnlineSecurity        0
 internet_OnlineBackup          0
 internet_DeviceProtection      0
 internet_TechSupport           0
 internet_StreamingTV           0
 internet_StreamingMovies       0
 account_Contract               0
 account_PaperlessBilling       0
 account_PaymentMethod          0
 account_Charges_Monthly        0
 account_Charges_Total         11
 dtype: int64,
 customerID                   0
 Churn                        0
 customer_gender              0
 customer_SeniorCitizen       0
 customer_Partner             0
 customer_Dependents          0
 customer_tenure              0
 phone_PhoneService           0
 phone_Multiple

In [85]:
df['customerID'].nunique()

7267

In [86]:
df.isin(['Yes', 'No']).all()

customerID                   False
Churn                        False
customer_gender              False
customer_SeniorCitizen       False
customer_Partner              True
customer_Dependents           True
customer_tenure              False
phone_PhoneService            True
phone_MultipleLines          False
internet_InternetService     False
internet_OnlineSecurity      False
internet_OnlineBackup        False
internet_DeviceProtection    False
internet_TechSupport         False
internet_StreamingTV         False
internet_StreamingMovies     False
account_Contract             False
account_PaperlessBilling      True
account_PaymentMethod        False
account_Charges_Monthly      False
account_Charges_Total        False
dtype: bool

In [87]:
df.isin([0, 1]).all()

customerID                   False
Churn                        False
customer_gender              False
customer_SeniorCitizen        True
customer_Partner             False
customer_Dependents          False
customer_tenure              False
phone_PhoneService           False
phone_MultipleLines          False
internet_InternetService     False
internet_OnlineSecurity      False
internet_OnlineBackup        False
internet_DeviceProtection    False
internet_TechSupport         False
internet_StreamingTV         False
internet_StreamingMovies     False
account_Contract             False
account_PaperlessBilling     False
account_PaymentMethod        False
account_Charges_Monthly      False
account_Charges_Total        False
dtype: bool

# Results

+ 224 empty values in `Churn`
+ 11 empty values in `account_Charges_total`
+ Wrong data type in `account_Charges_Total` > numeric values
+ No repeated customers
+ Columns with only 1/0 or Yes/No values > boolean
    * `customer_SeniorCitizen`
    * `customer_Partner`
    * `customer_Dependents`
    * `phone_PhoneService`
    * `account_PaperlessBilling`


In [ ]:
# Remove empty Churn rows



In [93]:
# Replace empty with 0 and turn column to floats
df['account_Charges_Total'] = df['account_Charges_Total'].apply(lambda x: '0' if x == '' else x)
df['account_Charges_Total'] = df['account_Charges_Total'].astype(float)

In [99]:
# Convert appropriate columns to booleans
df['customer_SeniorCitizen'] = df['customer_SeniorCitizen'].astype(bool)
cols_to_bool = ['customer_Partner', 'customer_Dependents', 'phone_PhoneService', 'account_PaperlessBilling']
df[cols_to_bool] = df[cols_to_bool].replace({'Yes' : True, 'No' : False}).astype('boolean')

In [100]:
df.sample(10)

customerID Churn customer_gender  customer_SeniorCitizen  \
2910  4039-HEUNW   Yes            Male                    True   
2768  3834-XUIFC   Yes            Male                   False   
5674  7753-USQYQ    No            Male                   False   
6610  9089-UOWJG    No          Female                   False   
6771  9348-ROUAI    No          Female                   False   
4954  6769-DCQLI   Yes            Male                   False   
6844  9443-JUBUO    No            Male                   False   
7183  9878-TNQGW   Yes            Male                   False   
7201  9907-SWKKF   Yes          Female                    True   
5576  7622-NXQZR    No            Male                   False   

      customer_Partner  customer_Dependents  customer_tenure  \
2910              True                False               36   
2768             False                False                7   
5674             False                False               55   
6610              True                 True               12   
6771              True                 True               59   
4954             False                False               52   
6844              True                 True               72   
7183             False                False                2   
7201             False                False                1   
5576             False                False               28   

      phone_PhoneService phone_MultipleLines internet_InternetService  ...  \
2910                True                 Yes              Fiber optic  ...   
2768                True                 Yes              Fiber optic  ...   
5674                True                  No                      DSL  ...   
6610                True                  No                      DSL  ...   
6771                True                 Yes                      DSL  ...   
4954                True                  No              Fiber optic  ...   
6844               False    No phone service                      DSL  ...   
7183                True                  No              Fiber optic  ...   
7201               False    No phone service                      DSL  ...   
5576               False    No phone service                      DSL  ...   

     internet_OnlineBackup internet_DeviceProtection internet_TechSupport  \
2910                   Yes                        No                   No   
2768                    No                        No                   No   
5674                   Yes                        No                  Yes   
6610                    No                       Yes                   No   
6771                   Yes                       Yes                  Yes   
4954                   Yes                       Yes                  Yes   
6844                   Yes                       Yes                  Yes   
7183                    No                        No                   No   
7201                    No                        No                   No   
5576                    No                        No                  Yes   

     internet_StreamingTV internet_StreamingMovies account_Contract  \
2910                  Yes                      Yes   Month-to-month   
2768                   No                      Yes   Month-to-month   
5674                   No                      Yes         One year   
6610                  Yes                       No   Month-to-month   
6771                  Yes                      Yes         Two year   
4954                  Yes                      Yes         One year   
6844                  Yes                      Yes         Two year   
7183                   No                      Yes   Month-to-month   
7201                   No                       No   Month-to-month   
5576                   No                       No   Month-to-month   

     account_PaperlessBilling      account_PaymentMethod  \
2910                     